### ISSUE TITLE
Extra KMS keys created when prepare stage fails

### ISSUE THREAD

**Issue by twdcnz** [awslabs/landing-zone-accelerator-on-aws#203]
**Describe the bug**
When the prepare pipeline stage fails during initial runs many extra KMS keys without aliases are created, but not deleted. When it runs again more keys are created. Each time the phase runs more KMS keys are created, but never removed.

**To Reproduce**
Create a new Control Tower landing zone.
Create a single AWS workload account in the landing zone using control tower.
Deploy LZA without adding the new workload account to the accounts-config.yaml file

**Expected behavior**
When the prepare stage fails first run it should delete any KMS keys it creates. This could be somewhat difficult to implement, and there may be some issues removing these keys in some situations I haven't considered.

**Please complete the following information about the solution:**

- [ ] Version: 1.4.2

To get the version of the solution, you can look at the description of the created CloudFormation stack. For example, "_(SO0021) - Video On Demand workflow with AWS Step Functions, MediaConvert, MediaPackage, S3, CloudFront and DynamoDB. Version **v5.0.0**_". If the description does not contain the version information, you can look at the mappings section of the template:

```yaml
Mappings:
  SourceCode:
    General:
      S3Bucket: "solutions"
      KeyPrefix: "video-on-demand-on-aws/v5.0.0"
```

- [ ] Region: ap-southeast-2
- [ ] Was the solution modified from the version published on this repository? No
- [ ] If the answer to the previous question was yes, are the changes available on GitHub?
- [ ] Have you checked your [service quotas](https://docs.aws.amazon.com/general/latest/gr/aws_service_limits.html) for the sevices this solution uses? N/A
- [ ] Were there any errors in the CloudWatch Logs? N/A

**Screenshots**
If applicable, add screenshots to help explain your problem (please **DO NOT include sensitive information**).

![image](https://github.com/awslabs/landing-zone-accelerator-on-aws/assets/91913779/9be78d7d-9cdc-4c90-b2bb-a2e205aaf2be)

**Additional context**

**Comment by bhkhatri221** [comment]
Hello @twdcnz, thank you for filing an issue with the Landing Zone Accelerator team! We have been able to replicate this behavior in an internal environment. I have created an issue for our team, and I will update this thread once I have an update on the team's progress.

Thank you for your interest in the Landing Zone Accelerator!

**Comment by thorben-akent** [comment]
We are also seeing this behavior a year later. @bhkhatri221 any updates on this?

**Comment by bhkhatri221** [comment]
Thank you @thorben-akent  for reaching out regarding the above bug report. We have an update for you. After careful consideration and planning, we've decided to address this issue as a feature enhancement in an upcoming release. The team will address this enhancement though improvements to the [CloudFormation enhances deletion policies](https://aws.amazon.com/about-aws/whats-new/2023/07/aws-cloudformation-deletion-policies-dev-test-cycle/). Until this enhancement is deployed, we recommend you review your existing KMS keys, and delete any solution-generated keys without aliases, but only after making sure they are no longer needed for decryption of logs or other artifacts.

Thank you for supporting Landing Zone Accelerator! I will keep this issue open in the meantime until this is released. Please let me know if you have any questions or concerns in the meantime."

**Comment by bpauwels** [comment]
Dear @bhkhatri221, is there an update on this issue? With which release are you planning to deliver this enhancement?

**Comment by nagmesh** [comment]
Thanks for raising this — the behavior is real, but it's intentional rather than a bug, and we owe better documentation around it. Let me explain the design and your options.

## Why LZA retains KMS keys across all stacks

LZA is a solution to support highly-regulated workloads and complex compliance requirements, and the KMS keys it provisions across every managed account encrypt customer data at rest — CloudWatch logs, S3 access logs, S3 buckets, Lambda environment variables, Session Manager logs, EBS volumes, AWS Backup vaults, and central audit logs in the LogArchive account. Some of that data falls under regulatory retention requirements that span years (commonly 7 for financial and healthcare workloads).

If we let CloudFormation discard a KMS key on rollback or stack deletion, any data already encrypted under that key would become permanently unreadable — including data customers may need to demonstrate compliance with later. That's a one-way door we won't take on customers' behalf.

So **all customer-data-bearing CMKs across LZA stacks are deliberately set to `RemovalPolicy.RETAIN`**. The behavior you observed in the prepare stage is the same principle applied to every stack and account: prepare, key, logging, security, operations, network — in management, log-archive, audit, and workload accounts alike. CloudFormation honors `RETAIN` even on initial-create rollback, which is what produces the alias-less, unattached keys.

## What this looks like to you

After any failed LZA deployment — or after intentionally deleting an LZA-managed account or stack — you may find KMS keys with:

- No alias (alias creation is in the same stack and rolled back with it)
- The `Accelerator` tag (or your configured prefix) and a creation timestamp matching the failed deployment
- "Pending deletion" not set — they're active and billable (~$1/month per key)

## Your options

1. **Leave them.** Safest. Subsequent successful deployments create fresh keys with their aliases attached; the orphans don't block anything. Cost is ~$1/key/month.

2. **Schedule deletion yourself**, after confirming nothing in your account encrypted data under the orphaned keys. Audit CloudTrail for `kms:GenerateDataKey` and `kms:Encrypt` calls against the key ID before scheduling deletion:
   ```
   aws kms schedule-key-deletion --key-id <id> --pending-window-in-days 30
   ```
   The 30-day window gives you time to cancel if you discover any data was encrypted with the key.

3. **Filter them out of cost reports** by tagging the orphans (e.g. `accelerator:status=orphan-cleanup`) and excluding that tag from your reporting.

## What we'll do

We'll update the [Key management](https://docs.aws.amazon.com/solutions/latest/landing-zone-accelerator-on-aws/key-management.html) page in the Implementation Guide to document this behavior across all stacks and accounts, what the orphan keys look like, and the safe cleanup procedure above. We'll keep this issue open until that documentation update lands. We're going to keep the `RETAIN` policy as the default — the data-loss exposure of the alternative is too significant for the compliance posture LZA is built around.

**Comment by nagmesh** [comment]
[Key management](https://docs.aws.amazon.com/solutions/latest/landing-zone-accelerator-on-aws/key-management.html) page has been updated
